game/domain/body.lua

```diff
@@ -397,22 +397,18 @@ function Body:removeAllArmor()
 end
 
 function Body:takeDamageFrom(amount, source)
-  local blocked = math.min(amount, self.armor)
-  self.armor = self.armor - blocked -- should never go negative
-  local dmg = math.max(0, amount - blocked)
+  local blocked = false
+  if RANDOM.generate(100) <= self:getBlockChance() then
+    amount = math.floor(amount/2)
+    blocked = true
+  end
+  local absorbed = math.min(amount, self.armor)
+  self.armor = self.armor - absorbed -- should never go negative
+  local dmg = math.max(0, amount - absorbed)
   self.damage = math.min(self:getMaxHP(), self.damage + dmg)
   self.killer = source:getId()
   self:triggerWidgets(TRIGGERS.ON_HIT)
-  return dmg
-  -- print damage formula info (uncomment for debugging)
-  --[[
-  local str = "%s is being attacked with %d damage!\n"
-              .. "> %s rolls %dd%d for %d defense points!\n"
-              .. "> %s takes %d in damage!\n"
-  local name = self:getSpec('name')
-  print(str:format(name, amount, name, self:getDEF(), self:getBaseDEF(),
-                   defroll, name, dmg))
-  --]]--
+  return dmg, blocked
```

Comment:
whenever possible, avoid returning multiple values. althought it's less efficient, returning a table with a field "damage" and a field "wasBlocked" is way more readable and maintainable

Reply:
Done